lib/galaxy/webapps/galaxy/controllers/user.py

```diff
@@ -510,21 +510,23 @@ def login( self, trans, refresh_frames=[], **kwd ):
 
     def __validate_login( self, trans, **kwd ):
         """Validates numerous cases that might happen during the login time."""
-        message = escape( kwd.get( 'message', '' ) )
         status = kwd.get( 'status', 'error' )
-        email = kwd.get( 'email', '' )
+        login = kwd.get( 'login', '' )
         password = kwd.get( 'password', '' )
-        username = kwd.get( 'username', '' )
         redirect = kwd.get( 'redirect', trans.request.referer ).strip()
         success = False
-        user = trans.sa_session.query( trans.app.model.User ).filter( trans.app.model.User.table.c.email == email ).first()
+        user = trans.sa_session.query( trans.app.model.User ).filter(or_(
+            trans.app.model.User.table.c.email == login,
+            trans.app.model.User.table.c.username == login
+        )).first()
```

Comment:
This ``first()`` is troublesome. I suspect it is a remnant of inconsistent DB on Main back when we did not enforce email to be unique nor valid. Username uniqueness is DB-enforced so it should not change anything but we might want to think about how to do this right. I like using ``.one()`` and catching exceptions from sqlalchemy (as in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/libraries.py#L39)
Other ideas?